In [10]:
#url = "https://docs.google.com/spreadsheets/d/1eZj7EOdGe16P5QWHSIIHhkclNAog13e7xF6fPMoXfEY/edit#gid=1897262175"
name = "main"

import pandas as pd
import urllib.parse

sheet_id = "1eZj7EOdGe16P5QWHSIIHhkclNAog13e7xF6fPMoXfEY"
sheet_name = "Buy_Insta_Accounts"
#sheet_name_encoded = urllib.parse.quote(sheet_name)
url = f"https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}"

In [14]:
fn = "Anthony MainSheet [current projects] - Buy_Insta_Accounts.tsv"
Buy_Insta_Accounts_df = pd.read_csv(fn, sep="\t")

## Load data

In [2]:
import os 
import pandas as pd

files = os.listdir("data")

df_women_event = pd.read_csv(os.path.join("data", files[0]), sep="\t")
df_buy_insta_accounts_add_on = pd.read_csv(os.path.join("data", files[1]), sep="\t")
df_buy_biz = pd.read_csv(os.path.join("data", files[2]), sep="\t")
df_buy_insta_accounts = pd.read_csv(os.path.join("data", files[3]), sep="\t")
df_women_event_mined = pd.read_csv(os.path.join("data", files[4]), sep="\t")

## EDA

In [85]:
df_buy_biz.media_count.describe()

count     143.000000
mean      405.426573
std       583.785926
min         0.000000
25%        54.000000
50%       166.000000
75%       481.500000
max      3813.000000
Name: media_count, dtype: float64

In [31]:
df_women_event_mined.columns

Index(['username', 'userid', 'scraped_from', 'media_count', 'full_name',
       'following_count', 'follower_count', 'first_name', 'external_url',
       'email', 'contact_phone_number', 'cityName', 'category', 'biography',
       'address_street', 'Big 5'],
      dtype='object')

In [38]:
df_buy_insta_accounts.columns

Index(['username', 'Openness', 'Conscientiousness', 'Extroversion',
       'Agreeableness', 'Neuroticism', 'Type', 'media_count', 'full_name',
       'following_count', 'follower_count', 'first_name', 'external_url',
       'engagement_percent', 'email', 'contact_phone_number', 'cityName',
       'category', 'biography', 'address_street', 'AvgLikes', 'AvgComments'],
      dtype='object')

In [50]:
df_buy_biz[df_buy_biz['Openness'].notna()]

username Openness Conscientiousness Extroversion  \
0               goodeatzco      Low              High         High   
2               a1republic     High              High         High   
3           keychainsocial     High              High         High   
4        thegiftshoppe.nft    Vague               Low          Low   
5       officialclubsocial      Low              High         High   
..                     ...      ...               ...          ...   
138           irose.social     High              High         High   
139               nattcity     High               Low         High   
140             jondennill     High              High         High   
141            igmodels_co      Low              High        Vague   
142  skystrategicmarketing     High              High         High   

    Agreeableness Neuroticism      Type  media_count                full_name  \
0           Vague       Vague      Blog         1693       #GoodEatzCo | Andy   
2            High         Low  Business           30            A1 REPUBLIC ™   
3            High         Low  Business          214    Keychain Social, Inc.   
4           Vague       Vague  Business            9          the gift shoppe   
5           Vague       Vague  Business            6              Club Social   
..            ...         ...       ...          ...                      ...   
138           Low         Low  Business          338                    IROSE   
139           Low         Low  Personal         1129                 nathalie   
140          High         Low  Personal          124         Jonathan Dennill   
141         Vague       Vague      Blog          458    IG Models Worldwide ®   
142          High         Low  Business          427  Sky Strategic Marketing   

     following_count  ...  engagement_percent  \
0                145  ...                0.14   
2                613  ...                0.53   
3                  0  ...                 NaN   
4                  5  ...                0.57   
5                  2  ...                 NaN   
..               ...  ...                 ...   
138              422  ...                0.30   
139              686  ...                0.06   
140              733  ...                0.49   
141              158  ...                0.01   
142             2133  ...                0.18   

                               email contact_phone_number  \
0               goodeatzco@gmail.com                  NaN   
2               a1republic@yahoo.com         1.949934e+10   
3            info@keychainsocial.com                  NaN   
4        hello@fromthegiftshoppe.com                  NaN   
5           clubsocialmgmt@gmail.com                  NaN   
..                               ...                  ...   
138           sharon@irosesocial.com                  NaN   
139  nathalia@nfluenceconsulting.com                  NaN   
140            jon.dennill@gmail.com                  NaN   
141            igmodels.co@gmail.com                  NaN   
142  tanya@skystrategicmarketing.com         1.813337e+10   

                      cityName             category  \
0      Los Angeles, California              Blogger   
2    Newport Beach, California  Social Media Agency   
3      Los Angeles, California                  NaN   
4                          NaN                  NaN   
5      Los Angeles, California  Social Media Agency   
..                         ...                  ...   
138    Los Angeles, California      Product/service   
139                        NaN                Actor   
140                        NaN                  NaN   
141    Los Angeles, California  Social Media Agency   
142             Tampa, Florida                  NaN   

                                             biography  \
0    🍴Professional Food Consumer🤘🏻😛 Tiktok 🎥: @good...   
2    The #1 organic IG growth service for influence...   
3    Digital Social Media Agency • Unlock Your 

In [52]:
len(df_buy_insta_accounts_add_on)

2734

In [59]:
df_women_event.columns

Index(['username', 'media_count', 'full_name', 'following_count',
       'follower_count', 'first_name', 'external_url', 'engagement_percent',
       'email', 'contact_phone_number', 'cityName', 'category', 'biography',
       'address_street', 'MostRecentPostDate', 'AvgLikes', 'AvgComments'],
      dtype='object')

## Meaning of the columns
https://developers.facebook.com/docs/instagram-api/reference/ig-user/
This URL has field defintion.
* biography - Profile bio text.
* followers_count - Total number of Instagram users following the user.
* following_count - Total number of Instagram users the user follows.
* media_count - Total number of IG Media published on the user.


In [6]:
set(df_women_event_mined.username)

0        hannahmariethemuse
1            refinedlash.co
2         studio_irene_rose
3              alena_smily9
4        sincerelyluna.xoxo
                ...        
11276                lianev
11277                   kuz
11278           tanamongeau
11279             lilyachty
11280             jenselter
Name: username, Length: 11281, dtype: object

In [12]:
df_women_event

username  media_count                      full_name  \
0             ariannyceleste         6039       Arianny Celeste UFC Mama   
1                 sarahstage         1501                    Sarah Stage   
2     beauty_nurse_elizabeth         1060           Elizabeth Sommers RN   
3                  carlyrbel          493                      CARLY BEL   
4              houdaschimmel          507                 Houda Schimmel   
...                      ...          ...                            ...   
1467            houghtonmade            1  Memorable and Detailed Events   
1468               violacomm            3           Viola Communications   
1469  brainstorm.media.group            0         Brainstorm Media Group   
1470     littlestudio_beauty           33           Little Studio_Beauty   
1471           adtop1percent            0   Ann Dee | LA ENTREPRENEUR LA   

      following_count  follower_count first_name  \
0                 997         3288148    Celeste   
1                1472         2086691      Sarah   
2                 383         1504557  Elizabeth   
3                 191         1035410      Carly   
4                3041          693036      Houda   
...               ...             ...        ...   
1467              166              21     Lauren   
1468                3              20      Viola   
1469               32              20  Katherine   
1470              146              12   Laetitia   
1471               42               7        Ann   

                          external_url  engagement_percent  \
0     http://www.ariannyexclusive.com/                0.06   
1         https://linktr.ee/Sarahstage                0.01   
2     http://www.elizabethsommers.com/                0.23   
3                 http://carlybel.com/                0.14   
4                                  NaN                 NaN   
...                                ...                 ...   
1467                               NaN               10.95   
1468   http://violacommunications.com/                6.83   
1469  http://brainstormmediagroup.com/                0.00   
1470       http://www.littlestudio.us/                4.03   
1471                               NaN                0.00   

                                   email contact_phone_number  \
0             arianny@ariannyceleste.com                  NaN   
1               booksarahstage@gmail.com                  NaN   
2                    sommersrn@gmail.com                  NaN   
3                carlybelsnaps@gmail.com                  NaN   
4               infos@shimmypetstore.com                  NaN   
...                                  ...                  ...   
1467             laurenjmadrid@gmail.com         +13234008726   
1468        info@violacommunications.com                  NaN   
1469  katherine@brainstormmediagroup.com                  NaN   
1470            laetitia@littlestudio.fr                  NaN   
1471             adtop1percent@gmail.com                  NaN   

                     cityName       category  \
0     Los Angeles, California  Public figure   
1                         NaN            NaN   
2     Los Angeles, California  Health/beauty   
3                         NaN   Entrepreneur   
4     Los Angeles, California   Entrepreneur   
...                       ...            ...   
1467  Los Angeles, California            NaN   
1468           Tampa, Florida            NaN   
1469  Los Angeles, California            NaN   
1470                      NaN            NaN   
1471                      NaN   Entrepreneur   

                                              biography address_street  \
0     Raidens Mom 👶🏽🐶🐶🧿 Founder @girlfriendbox @them...            NaN   
1     Boy Mom 👶🏻👦🏻 Wife 💃🏻 & Online Fitness Coach 💪🏼...            NaN   
2     Aesthetic RN-BSN •Injection Artist 🌹 •Natural ...            NaN   
3     One of a kind ✨ • @carlybelx • shop @carlyclub...            NaN   
4        

In [9]:
len(set(df_women_event_mined.username).intersection(set(df_women_event.username)))

3

In [10]:
len(set(df_buy_insta_accounts.username).intersection(set(df_buy_insta_accounts_add_on.username)))

0

## Call Instagram Business Discovery API

`df_buy_insta_accounts` has 94 rows with big 5 score, total 5535 rows

`df_buy_biz` total 143 rows, 129 rows have big 5 scores

`df_buy_insta_accounts_add_on` 2734 rows, 95 have big 5 scores

=======

`df_women_event_mined` Big 5 column is empty, total 11281 rows


`df_women_event` 1472 rows, no big five scores

In [3]:
import requests
import json
import pandas as pd
def get_media_metadata(ig_user_name, token, pages=1):
    params = {
        "ig_username": ig_user_name,
        "access_token": token,
        "endpoint_base":"https://graph.facebook.com/v17.0/",
        "instagram_account_id":"17841400224190654"
    }
    biz_discovery_first_page = '){username,ig_id,id,profile_picture_url,biography,follows_count,followers_count,media_count, media{comments_count, like_count, caption, timestamp, media_type}}' 

    endpointParams = dict() # parameter to send to the endpoint
    
    endpointParams['fields'] = 'business_discovery.username(' + params['ig_username'] + biz_discovery_first_page
    
    # string of fields to get back with the request for the account
    endpointParams['access_token'] = params['access_token'] # access token

    url = params['endpoint_base'] + params['instagram_account_id'] # endpoint url
    
    data = requests.get(url, endpointParams)
    if data.ok:
        dict_ = json.loads(data.content)
    else:
        return data.reason
    try:
        df = pd.DataFrame(dict_['business_discovery']['media']['data'])
    except:
        print(f"{ig_user_name} has bad data")
        return dict_
    
    pages -=1
    while pages>0:
        #d1 = json.loads(data.content)
        try:
            cursor = dict_['business_discovery']['media']['paging']['cursors']['after']
        except:
            return "no cursor"
        biz_discovery_following_page = f"){{username,website,name,ig_id,id,profile_picture_url,biography,follows_count,followers_count,media_count, media.after({cursor}){{comments_count, like_count, caption, timestamp, media_type}}}}"
        endpointParams['fields'] = 'business_discovery.username(' + params['ig_username'] + biz_discovery_following_page
        url = params['endpoint_base'] + params['instagram_account_id'] # endpoint url
        data = requests.get(url, endpointParams)
        dict_ = json.loads(data.content)
        df = pd.concat([df, pd.DataFrame(dict_['business_discovery']['media']['data'])])
                       
        pages -=1
        
    
    return df


In [7]:
from tqdm import tqdm
import os
from pathlib import Path

# go to: https://developers.facebook.com/tools/explorer/
# to get token
token ="EAACccrMmWa0BANGimnovFDWl55PRy2AWk5sZBuMTwgtZAxOMhiDTovrNiPKvB6E5Oq7ZBWRmlAa9ldTZCPOl7uAj3zmC8PFEnv7DP8paCGtPKIoVsQzCKnP9xOtFg3vi1AFLTvfq5rFbplT4yeVZCitoLcEzSzxDwSCXWw0fCEPGSalCgypaFosaMc7ZCEA7aYps6jGk1AnVxxjA3prZBTQ"

usernames = df_buy_insta_accounts[df_buy_insta_accounts['Openness'].notna()]['username']
folder = "buy_insta_accounts"
try:
    os.mkdir(f"feature/{folder}")
except:
    pass


scraped_users = []
for entry in Path(f"feature/{folder}").iterdir():
    # check if it a file
    if entry.is_file():
        scraped_users.append(entry.stem)

bad_users = []
for ig_username in tqdm(usernames.tolist()):
    if ig_username in scraped_users:
        continue

    res = get_media_metadata(ig_username, token, 2)
    if type(res) is str:
        if res == "Forbidden":
            print("possibly reach rate limit")
            break
        bad_users.append({ig_username: res})
    else:
        res.to_parquet(f"feature/{folder}/{ig_username}.parquet")

100%|██████████| 94/94 [00:30<00:00,  3.05it/s]


In [6]:
res

comments_count  like_count  \
0                0          17   
1                0          11   
2                0           6   
3                4          37   
4                1          16   
5                0          23   
6                0           4   
7                1          17   
8                3          22   
9                0           9   
10               0           7   
11               0           5   
12               0          18   
13               9          44   
14               1          25   
15               1          18   
16               0           8   
17               0          18   
18               1          12   
19               1          19   
20               0           5   
21               9          26   
22               5          19   
23               0          27   
24               0          14   
0                0           8   
1                2          32   
2                1          17   
3                0          19   
4                0          19   
5                2          16   
6                0          30   
7                0          15   
8                0           9   
9                0          13   
10               0           9   
11               0           6   
12               0          12   
13               1          10   
14               0          20   
15               0           9   
16               0          13   
17               0           9   
18               1          26   
19               1          14   
20               0          16   
21               3          41   
22               3          23   
23               1          26   
24               0          12   

                                              caption  \
0   🤣 SMILE BIG WITH FROSTING ALL OVER YOUR FACE I...   
1   Thank you to all the volunteers, sponsors, and...   
2   Thank you, Emma tampabaychamber_ emergingleade...   
3   We’re celebrating Sky Strategic Marketing toda...   
4   Meet nampurrie with cultivateadvisors! She is ...   
5   🦈 Just another FINtastic day at Sky Strategic ...   
6   May the 4th be with you! Happy Star Wars Day! ...   
7   ALL 👀's on our birthday girl, Audra Perry Butl...   
8   Wonderful afternoon celebrating Women's Histor...   
9   Awesome Working Women Of Tampa Bay Brunch Like...   
10  Starting this week with sunshine and kindness ...   
11  **ANOTHER Taste of South Tampa restaurant spot...   
12  Still thinking of those delicious🍦flavors OddF...   
13  STOP SCROLLING…and join us in celebrating Nati...   
14  You know your Thursday is going to be excellen...   
15  You've seen our "Ghosts of Christmas Past", he...   
16  Who are these "Ghosts of Christmas Past"? Can ...   
17  We had a blast creepin' it real delivering 🎃 a...   
18                  Happy Birthday Gabriel Guerrero!!   
19  Even though it rained cats and dogs all week, ...   
20  Happy Independence Day! 🇺🇸\n\n#HappyBirthdayAm...   
21  Celebrated the last day of #PrideMonth with th...   
22  🎈Happy Birthday to our Sporty Spice, kellyfm86...   
23  Never Say Never!\n\nThank you, Brian Ford, COO...   
24  In memory of many, in honor of all. 🇺🇸\n\n#Mem...   
0   "My mother is so full of joy and life. I am he...   
1   Today's office, rawr! 🦁🐯🐨🐵🐻🐧🦉🐴🐢🐅🦏🦛🐘🦧🦓🐆🦜🦚🦩🐖🐿🦥🦦🦡...   
2   Congratulations to Katie McGill and everyone a...   
3   Today's forecast - s🌞nshine, noms, and adult b...   
4   Working Women Of Tampa Bay Back To Business Br...   
5   #Repost sarinafazannews\n• • • • • •\nNewsom E...   
6   We had a blast visiting some of our favorite p...   
7   From our family to yours, Happy Easter!\n\n#Ha...   
8   Your life. 📜 Your story. \n✏️ Write well. Edit...   
9   🎈Happy Birthday to you🎈\n🎈Happy Birthday to yo...   
10  "You can't use up creativity.\nThe more you us...   
11  "The ones who are crazy enough to think that t...   
12  Check out a few behind the scenes shots with o...   
13  The sky is the limit for digital broadcast

In [17]:
df_buy_insta_accounts_add_on[df_buy_insta_accounts_add_on['Openness'].notna()]

username Openness Conscientiousness Extroversion  \
0       patrickbreenmusic      Low              High          Low   
1    love.lindsey.wheeler     High              High         High   
2            authorkwells     High               Low         High   
3       aleksatodorovicev     High              High         High   
4        grigoruadisveley     High              High         High   
..                    ...      ...               ...          ...   
95            dal.vivanco      Low              High         High   
96          musicallygina     High              High         High   
97            sammy_finch     High              High         High   
98  hopetaylorphotography     High              High         High   
99           coach_moyses      Low              High         High   

   Agreeableness Neuroticism      Type  media_count  \
0           High         Low  Personal          504   
1           High         Low  Personal         2381   
2           High         Low  Personal          129   
3            Low         Low  Personal          897   
4            Low         Low  Personal        10742   
..           ...         ...       ...          ...   
95          High         Low  Personal          199   
96          High         Low  Personal          349   
97          High         Low  Personal          586   
98          High         Low  Business         5744   
99           Low         Low  Personal          900   

                         full_name  following_count  ...  \
0                    Patrick Breen              893  ...   
1                  LINDSEY WHEELER             1332  ...   
2    Author Kamisha Wells RDKS CEO              314  ...   
3                Todorović Aleksa              876  ...   
4      Anzhelika Grigoryadis Veley              980  ...   
..                             ...              ...  ...   
95             Hell’s Lady Barber🔥              121  ...   
96                          Gina 🎶              826  ...   
97                     Sammy Finch              770  ...   
98                     Hope Taylor             1436  ...   
99  M o y s e s  S a n t a n a🇧🇷🇺🇸             1352  ...   

                                         external_url engagement_percent  \
0                                                 NaN              2.793   
1                        https://www.lovelindsey.com/              1.063   
2            https://www.amazon.com/shop/authorkwells              1.804   
3                 https://linktr.ee/aleksatodorovicev              0.703   
4                 https://paysend.com/referral/01g6gi              0.470   
..                                                ...                ...   
95                                                NaN              2.330   
96                    https://linktr.ee/musicallygina              0.330   
97                        http://linktr.ee/getcontact              0.787   
98            https://linktr.ee/hopetaylorphotography              1.765   
99  https://www.evogennutrition.com/?aff=1777&_utm...              1.239   

                                email  contact_phone_number  \
0       patrick@patrickbreenmusic.com                   NaN   
1           lindseyslularoe@gmail.com          +17576206307   
2              kamishawells@yahoo.com          +15627350110   
3           todorovicaskela@gmail.com                   NaN   
4    anzhelika-grigorjadis@rambler.ru                   NaN   
..                                ...                   ...   
95               daaleery@hotmail.com                   NaN   
96           ginareneemusic@gmail.com                   NaN   
97              sammyfinch8@gmail.com                   NaN   
98  hopetaylorphotography@hotmail.com                   NaN   
99              moyses_1979@yahoo.com                   NaN   

                   cityName         category  \
0                       NaN         Musician   
1                       NaN     Entrepre

In [50]:
bad_users

comments_count  like_count  \
0               10         177   
1                9         459   
2                4         334   
3                2         141   
4               16         548   
5               13         741   
6               35        1791   
7               24         321   
8                4          94   
9               32         603   
10              22         524   
11              93         695   
12               3         187   
13              20         230   
14              18         684   
15              16         265   
16              15         284   
17               8         261   
18              26         262   
19              99       11972   
20              60         662   
21              12         262   
22               8         230   
23              61         678   
24              12         223   
0                4         378   
1               26         620   
2              142         877   
3               71         569   
4               14         228   
5                7         175   
6                4         269   
7               15         338   
8                9         431   
9               42         537   
10              10         208   
11              26         525   
12              10         282   
13              36        1025   
14              22         277   
15              28         629   
16              22         148   
17               4        1274   
18               9         253   
19               8         321   
20              26         502   
21              31         527   
22              16         740   
23               0          29   
24               4          30   

                                              caption  \
0   Disfruta experiencias, consigue energías única...   
1   ¿Porque no buscar la profesionalización?. Si r...   
2   ¿Es lo mismo vender $1,000 USD en productos qu...   
3   Mis pensamientos 💭 #24\n\n✅ ¿Cuánto vale mi li...   
4   Uno de mis grandes errores y de tanto que me l...   
5   ¿Que está haciendo para cambiar tu realidad? |...   
6   El cambio de mentalidad 🧠 es necesario y este ...   
7   Describo los 4 tipos de usuarios de las redes ...   
8   Mis Pensamientos 💭 #23\n\n✅ ¿Todos podemos ser...   
9   Salí por las calles de Midtown Miami y me veía...   
10  ¿Que harías si hoy tienes $1,000,000 usd?. No ...   
11  ¿Es necesario mostrar LifeStyle, dinero , etc?...   
12  ¿Y tú qué contenido consumes? Después no te qu...   
13  Mis pensamientos 💭 #22\n\n✅ contenido viral 💣💣...   
14  Feliz día a todas las madres.\n\nEl mejor prem...   
15  ¿Quien hace viral el contenido? Este video no ...   
16  Mis Pensamientos 💭 #21\n\n✅ #ayahusca vive la ...   
17  La edad no te impide empezar de 0 y mucho meno...   
18               Mis pensamientos 💭 #20 \n\n✅ Actitud   
19  El trabajo no denigra, hazlo con amor y ten ac...   
20  Yo ✅ siendo Yo y si no puedes ser tú ¿para qué...   
21  Preocúpate por ti, cuando seas tu prioridad, d...   
22  ✅ ¿Monetizar Online? \nEn el episodio #18 habl...   
23  Que no te de miedo empezar| Sígueme en gocholi...   
24  Trabaja tus capacidades si no las tienes créal...   
0   ¿Construyendo o sobreviviendo? Despiértate 🚀| ...   
1   Y tus fines de semana son de construir la vida...   
2   Prefiero estar en depresión con abundancia, qu...   
3   Siguen vendiéndonos la pobreza te quieren dorm...   
4   Paso a paso para hacer Trading y no morir en e...   
5   Sube volumen solo 1 minuto Cierra los ojos y e...   
6   Mis pensamientos 💭 #17\n\n✅ ¿Como crear capaci...   
7   Podría cambiar tu vida de manera exponencial ☀...   
8   ¿Promueves mentalidad de crecimiento y atraes ...   
9   ¿Conoces alguna? Es de Energías pero pocos ent...   
10  Como abrir tu cuenta en MOON4TRADERS BROKERAGE...   
11  🤭🤭 Mujer, chica, novia, esposa comparte este R...   
12  Mis pensamientos 💭 #16\n\n✅ Misión de todos lo...   
13  El entorno ⭕️ ¿como es el tuyo? ¿Es duro ?

In [14]:
len(os.listdir("feature/buy_insta_accounts"))

61

In [53]:
df_buy_insta_accounts[df_buy_insta_accounts.username == ig_username]

username Openness Conscientiousness Extroversion Agreeableness  \
5452  anyelo_rr     High               Low         High           Low   

     Neuroticism      Type  media_count  full_name  following_count  ...  \
5452         Low  Personal          898  Anyelo RR              350  ...   

                                   external_url engagement_percent  \
5452  https://m.youtube.com/watch?v=GqGVkJVk6Ds              2.037   

                           email  contact_phone_number cityName category  \
5452  anyelorrtheghost@gmail.com                   NaN      NaN   Artist   

                                              biography address_street  \
5452  👻 GHOST 👻  BOOKING : Anyelorrbooking@gmail.com...            NaN   

     AvgLikes AvgComments  
5452   988.39       39.67  

[1 rows x 22 columns]

In [37]:
!pip install emot

[PosixPath('feature/buy_insta_accounts'),
 PosixPath('feature/buy_insta_accounts_add_on'),
 PosixPath('feature/buy_biz')]

In [15]:
#combine 
import pandas as pd
from pathlib import Path

p = Path('feature')
folders = [x for x in p.iterdir() if x.is_dir() and x.stem.startswith("buy")]

full_df = pd.DataFrame()
for data_dir in folders:
    for parquet_file in data_dir.glob('*.parquet'):
        ig_username = parquet_file.stem
        df = pd.read_parquet(parquet_file)
        df['ig_username'] = ig_username
        full_df = pd.concat([full_df, df])

In [16]:
full_df

comments_count  like_count  \
0               22       464.0   
1                7       130.0   
2               36       846.0   
3               13       190.0   
4               23       460.0   
..             ...         ...   
20              23       842.0   
21              14       627.0   
22              10       692.0   
23              20       586.0   
24              43       742.0   

                                              caption  \
0                             Baby Rock on tour 🤘🏽🍼🎶🚐   
1   Este viernes actuaré con mi banda en  Mislata ...   
2   Resumen de la minigira Aragonesa. \nLo pasamos...   
3   Esta semana nos vamos de conciertos por tierra...   
4   ¡Gracias, MADRID! 🥰👏🏼\n Siempre es un gustazo ...   
..                                                ...   
20  “The desire to create is one of the deepest ye...   
21  "It is presentation which lifts the card trick...   
22  "Creativity makes a leap, then looks to see wh...   
23                When nothing goes right... Go left!   
24  “Creativity is seeing what everyone else has s...   

                   timestamp      media_type                 id  \
0   2023-06-02T12:37:44+0000           IMAGE  17967961775389122   
1   2023-05-30T12:01:51+0000           IMAGE  17960665451581426   
2   2023-05-29T15:02:14+0000  CAROUSEL_ALBUM  18362698420034400   
3   2023-05-22T11:01:23+0000  CAROUSEL_ALBUM  17956348343561063   
4   2023-05-15T17:18:16+0000  CAROUSEL_ALBUM  17977392881476269   
..                       ...             ...                ...   
20  2015-10-24T21:25:32+0000           IMAGE  17845192774030704   
21  2015-10-21T06:36:54+0000           IMAGE  17845122445030704   
22  2015-10-20T08:20:54+0000           IMAGE  17844830602030704   
23  2015-10-18T20:14:23+0000           IMAGE  17844269023030704   
24  2015-10-17T17:51:59+0000           IMAGE  17844249061030704   

       ig_username  
0   georginamusica  
1   georginamusica  
2   georginamusica  
3   georginamusica  
4   georginamusica  
..             ...  
20  keychainsocial  
21  keychainsocial  
22  keychainsocial  
23  keychainsocial  
24  keychainsocial  

[13317 rows x 7 columns]

## Preprocess Features

In [22]:
from emot.emo_unicode import UNICODE_EMO, EMOTICONS
import re

def lower_text(text):
    try:
        text = text.lower()
    except:
        return ""
    return text
def remove_newline(string):
    string = string.replace("\n", "")
    return string
def remove_punctuation(string): 
    punctuations = '''()-[]{};:'"\<>/@#$%^&*~'''
    for x in string.lower(): 
        if x in punctuations: 
            string = string.replace(x, "")
        elif x == ",._":
            string = string.replace(x, " ")
        elif x == "?":
            string = string.replace(x, "[question mark]")
    return string

def convert_emojis(text):
    for emot in UNICODE_EMO:
        text = text.replace(emot, "_".join(UNICODE_EMO[emot].replace(",","").replace(":","").split()))
    return text

def convert_emoticons(text):
    for emot in EMOTICONS:
        text = re.sub(u'('+emot+')', "_".join(EMOTICONS[emot].replace(",","").split()), text)
    return text

def combine_metadata(row):
    return f"caption {row['caption']}, is of type {row['media_type'].lower()}, has {row['comments_count']} comments and {row['like_count']} likes"


In [33]:
full_df['full_caption'] = full_df.apply(combine_metadata, axis=1)
full_df['full_caption'] = full_df['full_caption'].apply(lower_text)
full_df['full_caption'] = full_df['full_caption'].apply(remove_newline)
full_df['full_caption'] = full_df['full_caption'].apply(convert_emojis)
full_df['full_caption'] = full_df['full_caption'].apply(convert_emoticons)
full_df_concat = full_df[['ig_username','full_caption']].groupby("ig_username").sum()

In [36]:
full_df_concat

full_caption
ig_username                                                        
36mktg            caption did you know henry ford made the 8 hou...
6staroileaters    caption so clean smiling_face_with_heart-eyesr...
740project        caption through the night crescent_moon • maet...
_boostfruit.com_  caption take your social media accounts out of...
_cloutfinder      caption appreciate you excellencealways folded...
...                                                             ...
wipmarketing      caption none, is of type image, has 1 comments...
xcarmeeen         caption enjoy this video while i disappear for...
xflvipers         caption power back.#fanggang | xfl, is of type...
yogabody_uk       caption summeeeeer! all the hard work of clean...
zgroupla          caption now booking grammy award winning artis...

[268 rows x 1 columns]

In [48]:
columns = ['username', 'Openness','Conscientiousness','Extroversion','Agreeableness','Neuroticism']
feature_df =full_df_concat.merge(gc_df, left_index=True, right_on="username", how="left")
gc_df = pd.concat([
    df_buy_insta_accounts[columns],
    df_buy_biz[columns],
    df_buy_insta_accounts_add_on[columns]
])

In [51]:
feature_df

full_caption          username  \
59    caption did you know henry ford made the 8 hou...            36mktg   
69    caption so clean smiling_face_with_heart-eyesr...    6staroileaters   
26    caption through the night crescent_moon • maet...        740project   
57    caption take your social media accounts out of...  _boostfruit.com_   
10    caption appreciate you excellencealways folded...      _cloutfinder   
...                                                 ...               ...   
71    caption none, is of type image, has 1 comments...      wipmarketing   
14    caption enjoy this video while i disappear for...         xcarmeeen   
5461  caption power back.#fanggang | xfl, is of type...         xflvipers   
67    caption summeeeeer! all the hard work of clean...       yogabody_uk   
84    caption now booking grammy award winning artis...          zgroupla   

      Openness  Conscientiousness  Extroversion  Agreeableness  Neuroticism  
59         1.0                1.0           1.0            1.0          0.0  
69         0.0                1.0           0.5            0.5          0.5  
26         1.0                0.0           0.0            0.0          0.5  
57         1.0                0.0           1.0            0.5          0.5  
10         1.0                0.0           1.0            0.5          0.5  
...        ...                ...           ...            ...          ...  
71         1.0                1.0           1.0            1.0          0.0  
14         0.0                0.0           1.0            0.0          1.0  
5461       0.0                1.0           1.0            1.0          0.0  
67         1.0                1.0           1.0            1.0          0.0  
84         1.0                0.0           1.0            0.5          0.5  

[269 rows x 7 columns]

In [50]:
feature_df['Openness'] = feature_df['Openness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Conscientiousness'] = feature_df['Conscientiousness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Extroversion'] = feature_df['Extroversion'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Agreeableness'] = feature_df['Agreeableness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Neuroticism'] = feature_df['Neuroticism'].map({"High":1.0,"Low":0,"Vague":0.5})

In [94]:
feature_so_far = feature_so_far.dropna()

In [52]:
feature_df.to_parquet("feature.parquet")